### PR TITLE
Update openvino api to 2023.2.0

### PR DESCRIPTION
**Pull request by xiaoyu-work**
## Describe your changes

Update openvino api to 2023.2.0

## Checklist before requesting a review
- [x] Add unit tests for this change.
- [x] Make sure all tests can pass.
- [x] Update documents if necessary.
- [x] Lint and apply fixes to your code by running `lintrunner -a`
- [ ] Is this a user-facing change? If yes, give a description of this change to be included in the release notes.

### CODE REVIEW

File: olive/passes/pass_config.py [hunk 1]
```diff
@@ -38,6 +40,7 @@ class PassConfigParam(ConfigParam):
     """
 
     searchable_values: SearchParameter = None
+    allowed_values: Set = None
```
Comment (guotuofeng) [inline olive/passes/pass_config.py:43]: what's the field for?

Comment (guotuofeng) [reply olive/passes/pass_config.py:43]: I think it is a user-facing change?

Comment (xiaoyu-work) [reply olive/passes/pass_config.py:43]: This is used to validate param value. It's not a user facing change as this validation is internal only.

Comment (guotuofeng) [reply olive/passes/pass_config.py:43]: PassConfigParam is used as user-facing stuff.

File: olive/passes/openvino/quantization.py [hunk 2]
```diff
@@ -65,39 +82,81 @@ def _default_config(accelerator_spec: AcceleratorSpec) -> Dict[str, PassConfigPa
                 type_=Union[DataConfig, Dict],
                 description="Data config for calibration, required if dataloader_func is None.",
             ),
-            "metric_func": PassConfigParam(
-                type_=Union[Callable, str],
+            "model_type": PassConfigParam(
+                type_=str,
```
Comment (guotuofeng) [inline olive/passes/openvino/quantization.py:86]: should enum be better than allowed_values?

File: olive/passes/pass_config.py (hunk 1 above)
Comment (xiaoyu-work) [reply olive/passes/pass_config.py:43]: I don't think so. `PassConfigParam` is defined by developer to indicate the parameter's attributes. As for users, they only need to provide a dict.

File: olive/passes/openvino/quantization.py (hunk 2 above)
Comment (xiaoyu-work) [reply olive/passes/openvino/quantization.py:86]: Do you mean defining allowed values as enum? `allowed_values` set is simpler than an enum with additional definition and validation.

File: olive/passes/openvino/conversion.py [hunk 3]
```diff
@@ -19,31 +19,47 @@ class OpenVINOConversion(Pass):
...
-                    "Input can be set by passing a list of tuples. "
-                    "Each tuple should contain input name and optionally input type or input shape."
+                    "Set or override shapes for model inputs. "
+                    "It configures dynamic and static dimensions in model inputs "
+                    "depending on your inference requirements."
                 ),
             ),
-            "input_shape": PassConfigParam(
-                type_=List[int],
+            "example_input": PassConfigParam(
```
Comment (jambayk) [inline olive/passes/openvino/conversion.py:30]: since torch tensor or other inputs cannot be serialized, should this be made into a function or string name paramater instead. like for `dummy_input_func` or `dataloader_func`

File: olive/passes/openvino/conversion.py [hunk 4]
```diff
@@ -19,31 +19,47 @@ class OpenVINOConversion(Pass):
...
+                    "https://docs.openvino.ai/2023.2/openvino_docs_OV_Converter_UG_Conversion_Options.html"
+                ),
+            ),
+            "output_model ": PassConfigParam(
+                type_=str,
+                default_value="ov_model",
+                required=False,
+                description=(
+                    "Name of the output OpenVINO model. "
+                    "The output model will be saved to the same directory as the input model."
```
Comment (jambayk) [inline olive/passes/openvino/conversion.py:62]: does look like this is the case based on the code? 

Also maybe a name like `output_file_name` or something would be clearer?

File: olive/passes/pass_config.py [hunk 5]
```diff
@@ -4,7 +4,9 @@
 # --------------------------------------------------------------------------
 from enum import Enum
 from pathlib import Path
-from typing import Callable, Dict, Optional, Type, Union
+from typing import Callable, Dict, Optional, Set, Type, Union
+
+from pydantic import Field
```
Comment (jambayk) [inline olive/passes/pass_config.py:9]: this needs to be imported from `olive.common.pydantic_v1` like in line 12 so that we get the correct `Field` from v1 api.

File: olive/passes/openvino/conversion.py (hunk 4 above)
Comment (xiaoyu-work) [reply olive/passes/openvino/conversion.py:62]: This is `output_model` name is same as OpenVINO api parameter name: https://docs.openvino.ai/2023.2/openvino_docs_OV_Converter_UG_Conversion_Options.html, to keep the api parameter names unified. But I forgot to update the description here. Will update it.

File: olive/passes/openvino/quantization.py (hunk 2 above)
Comment (guotuofeng) [reply olive/passes/openvino/quantization.py:86]: with enum you can only need definition. @jambayk, what's your thought?

Comment (guotuofeng) [reply olive/passes/openvino/quantization.py:86]: https://docs.pydantic.dev/2.0/usage/types/enums/

Comment (xiaoyu-work) [reply olive/passes/openvino/quantization.py:86]: Ah not aware that pydantic supports Enums here. This looks better than manual validation. Will update with Enum.

**Comment by xiaoyu-work** [comment]
There is a bug for openvino example. I removed it from pipeline for now to unblock this change. Will add it back once the bug is fixed.

File: olive/passes/pass_config.py [hunk 6]
```diff
@@ -38,6 +38,7 @@ class PassConfigParam(ConfigParam):
     """
 
     searchable_values: SearchParameter = None
+    allowed_values: Set = None
```
Comment (guotuofeng) [inline olive/passes/pass_config.py:41]: it seems this field could be removed?

File: .azure_pipelines/olive-ci.yaml [hunk 7]
```diff
@@ -119,9 +116,6 @@ jobs:
       bert_ptq_cpu_aml:
         exampleFolder: bert
         exampleName: bert_ptq_cpu_aml
-      cifar10_openvino_intel_hw:
-        exampleFolder: cifar10_openvino_intel_hw
-        exampleName: cifar10_openvino_intel_hw
```
Comment (guotuofeng) [inline .azure_pipelines/olive-ci.yaml:124]: do we need enable the new test?

File: olive/evaluator/olive_evaluator.py [hunk 8]
```diff
@@ -823,8 +822,7 @@ def _inference(
         device: Device = Device.CPU,
         execution_providers: Union[str, List[str]] = None,
     ) -> Tuple[OliveModelOutput, Any]:
-        session = model.prepare_session(inference_settings=self.get_inference_settings(metric), device=device)
-
+        session = model.prepare_session()
```
Comment (guotuofeng) [inline olive/evaluator/olive_evaluator.py:825]: >prepare_session [](http://example.com/codeflow?start=24&length=15)

why do we remove the device from the prepare_session?

File: olive/passes/pass_config.py (hunk 6 above)
Comment (xiaoyu-work) [reply olive/passes/pass_config.py:41]: Good catch. Will remove it.

File: olive/evaluator/olive_evaluator.py (hunk 8 above)
Comment (xiaoyu-work) [reply olive/evaluator/olive_evaluator.py:825]: We enable session device here: https://github.com/microsoft/Olive/blob/main/olive/evaluator/olive_evaluator.py#L833. We could refactor this to move this to `prepare_session`

File: olive/evaluator/olive_evaluator.py [hunk 9]
```diff
@@ -1017,14 +1015,13 @@ def _inference(
...
+        for input_data, label in dataloader:
+            session.infer({0: input_data})
+            result = session.get_output_tensor(0).data
             outputs = post_func(result) if post_func else result
-            if not isinstance(labels, list):
-                labels = [labels]  # noqa: PLW2901
-            preds.extend(outputs)
-            targets.extend(labels)
-            logits.extend(result)
+            preds.append(outputs)
```
Comment (jambayk) [inline olive/evaluator/olive_evaluator.py:1022]: is there a reason this is `append` and not `extend`. Is the data not in batches for openvino? Just one sample per iteration of the loop?

File: olive/passes/openvino/conversion.py [hunk 10]
```diff
@@ -15,37 +15,52 @@
...
+                    "For PyTorch it can be torch.Tensor. "
+                    "For Tensorflow it can be tf.Tensor or numpy.ndarray. "
                 ),
             ),
-            "extra_config": PassConfigParam(
+            "compress_to_fp16": PassConfigParam(
+                type_=bool,
+                default_value=True,
+                required=False,
+                description=("Compress weights in output OpenVINO model to FP16. Default is True."),
```
Comment (jambayk) [inline olive/passes/openvino/conversion.py:45]: nit: the parenthesis here are redundant?

File: olive/passes/openvino/conversion.py [hunk 11]
```diff
@@ -55,38 +70,51 @@ def _run_for_config(
...
-            model_input_param = "saved_model_dir"
         if model.framework == Framework.PYTORCH:
             input_model = model.load_model()
 
-        if config.get("extra_config") is None:
-            config["extra_config"] = {}
+        example_input = None
+        if config.get("example_input_func"):
+            example_input_func = config["example_input_func"]
+            if isinstance(example_input_func, str):
```
Comment (jambayk) [inline olive/passes/openvino/conversion.py:88]: nit: There is no need for the condition here. We can always directly use `example_input = self._user_module_loader.call_object(example_input_func)`. The `UserModuleLoader` object can call a function directly too https://github.com/microsoft/Olive/blob/main/olive/common/user_module_loader.py#L26

File: olive/passes/openvino/conversion.py [hunk 12]
```diff
@@ -15,37 +15,52 @@
...
     """Converts PyTorch, ONNX or TensorFlow Model to OpenVino Model."""
 
+    _requires_user_script = True
+
     @staticmethod
     def _default_config(accelerator_spec: AcceleratorSpec) -> Dict[str, PassConfigParam]:
         return {
             "input": PassConfigParam(
-                type_=List[Tuple],
+                type_=Union[Callable, str, List],
```
Comment (jambayk) [inline olive/passes/openvino/conversion.py:24]: not an issue but should we also make this type only as function or function name like we do in other places in Olive. 
The `List` case is only useful when someone want to use the python api but we don't have any use case for that except for the unit tests.

File: olive/evaluator/olive_evaluator.py (hunk 9 above)
Comment (xiaoyu-work) [reply olive/evaluator/olive_evaluator.py:1022]: Ah good catch! I made this change to test torchmetrics accuracy api feature but forgot to change it back. Will update this.

File: olive/passes/openvino/conversion.py (hunk 11 above)
Comment (xiaoyu-work) [reply olive/passes/openvino/conversion.py:88]: Will do

File: olive/passes/openvino/conversion.py (hunk 12 above)
Comment (xiaoyu-work) [reply olive/passes/openvino/conversion.py:24]: This is useful for those simpler model like our current example (even I removed this): https://github.com/microsoft/Olive/blob/main/examples/cifar10_openvino_intel_hw/config.json#L55

Comment (jambayk) [reply olive/passes/openvino/conversion.py:24]: Oh so, we can still provide the shapes as input?

Comment (xiaoyu-work) [reply olive/passes/openvino/conversion.py:24]: Yes, this is similar as onnx conversion, either input_shape (input) or example_input is needed.

File: test/unit_test/passes/openvino/test_openvino_conversion.py [hunk 13]
```diff
@@ -13,7 +14,7 @@ def test_openvino_conversion_pass(tmp_path):
     # setup
     input_model = get_pytorch_model()
     dummy_input = get_pytorch_model_dummy_input(input_model)
-    openvino_conversion_config = {"extra_config": {"example_input": dummy_input}}
+    openvino_conversion_config = {"example_input": dummy_input}
```
Comment (jambayk) [inline test/unit_test/passes/openvino/test_openvino_conversion.py:17]: this got changed to `example_input_func`?

File: olive/passes/openvino/conversion.py (hunk 12 above)
Comment (jambayk) [reply olive/passes/openvino/conversion.py:24]: My bad, this is still for input shapes. The name `input` was confusing but I guess, that's what the openvino api calls it too.

Comment (xiaoyu-work) [reply olive/passes/openvino/conversion.py:24]: Yeah this is how openvino api calls it. This is so confusing.

File: olive/passes/openvino/conversion.py [hunk 14]
```diff
@@ -55,38 +70,45 @@ def _run_for_config(
...
-        if model.framework == Framework.TENSORFLOW and Path(model.model_path).is_dir():
-            model_input_param = "saved_model_dir"
         if model.framework == Framework.PYTORCH:
             input_model = model.load_model()
 
-        if config.get("extra_config") is None:
-            config["extra_config"] = {}
+        example_input = None
+        if config.get("example_input_func"):
+            example_input = self._user_module_loader.call_object(config["example_input_func"])
```
Comment (jambayk) [inline olive/passes/openvino/conversion.py:87]: Could it be better if we called the `example_input_func` with the `input_model` object? Like how we do for the pytorch dummy input func. That way the input func can choose to use information from the input model like `model_attributed` if it wants. Helps when we don't want to hard code values into the user scripts to make it more flexible.

Comment (jambayk) [reply olive/passes/openvino/conversion.py:87]: in the ut, we could then directly use `get_pytorch_model_dummy_input` as the `example_input_func` directly.

Comment (xiaoyu-work) [reply olive/passes/openvino/conversion.py:87]: I don't see any use case of this. Any example for this case?

Comment (jambayk) [reply olive/passes/openvino/conversion.py:87]: yeah 
https://github.com/microsoft/Olive/blob/main/examples/llama2/user_script.py#L22
https://github.com/microsoft/Olive/blob/main/examples/whisper/code/user_script.py#L145

Comment (xiaoyu-work) [reply olive/passes/openvino/conversion.py:87]: (Not remember if our team had such discussion before) I still don't want to encourage user to use OliveModelHandler in the user script (or any Olive internal implementation), as they have to have pre-knowledge of how OliveModelHandler looks like and how it works. For the llama2 example and whisper example, I can see it is possible to avoid using PyTorchModelHandler but those codes can still work after refactoring.

Comment (xiaoyu-work) [reply olive/passes/openvino/conversion.py:87]: If we do want to use some metadata from OliveModel, I prefer to specially expose a attribute like "metadata" or "extra_data" to user. So they don't need to read our source code of model handler but they can know they can get some data from this attribute from Olive model.

Comment (jambayk) [reply olive/passes/openvino/conversion.py:87]: Sure, we can discuss about it separately and just have this pass keep its current signature. 

The `model_attributes` are mean to function similar to the "metadata" you mention. We could also do something like allow the user to provide `func_kwargs` like we do for metrics https://github.com/microsoft/Olive/blob/bc8b3312251484986972cf53ac8d4b2f4ed2ca1f/examples/notebooks/llama2/config.json#L91

File: test/unit_test/passes/openvino/test_openvino_conversion.py [hunk 15]
```diff
@@ -13,7 +14,7 @@ def test_openvino_conversion_pass(tmp_path):
     # setup
     input_model = get_pytorch_model()
     dummy_input = get_pytorch_model_dummy_input(input_model)
-    openvino_conversion_config = {"extra_config": {"example_input": dummy_input}}
+    openvino_conversion_config = {"example_input_func": dummy_input}
```
Comment (jambayk) [inline test/unit_test/passes/openvino/test_openvino_conversion.py:17]: This will fail because it's not a function.

File: olive/passes/openvino/conversion.py (hunk 14 above)
Comment (xiaoyu-work) [reply olive/passes/openvino/conversion.py:87]: Yeah, we can especially expose some model_attributes for all our models.

**Review by jambayk — APPROVED** [review]

**Review by guotuofeng — APPROVED** [review]
